File: src/main/java/net/digital_alexandria/R__/parser/Parser.java
package net.digital_alexandria.r__.parser;

import net.digital_alexandria.r__.exceptions.ParsingException;
import net.digital_alexandria.r__.lexer.Lexer;
import net.digital_alexandria.r__.lexer.Token;
import net.digital_alexandria.r__.lexer.TokenCategory;

/**
 * @author Simon Dirmeier {@literal mail@simon-dirmeier.net}
 */
public final class Parser
{
    private final Lexer _lexer;
    private Token _currToken;

    private Parser(final Lexer lexer)
    {
        this._lexer = lexer;
    }

    public static Parser instance(final Lexer lexer)
    {
        return new Parser(lexer);
    }

    public void init(final String text)
    {
        _lexer.init(text);
        _currToken = _lexer.next();
    }

    public AbstractSyntaxSubtree parse()
    {
        return expression();
    }

    @SuppressWarnings("unchecked")
    private AbstractSyntaxSubtree expression()
    {
        AbstractSyntaxSubtree root = term();

        while (_currToken.category() == TokenCategory.PLUS ||
               _currToken.category() == TokenCategory.MINUS)
        {
            Token<?> t = _currToken;
            switch (t.category())
            {
                case PLUS:
                    eat(TokenCategory.PLUS);
                    break;
                case MINUS:
                    eat(TokenCategory.MINUS);
                    break;
                default:
                    throw new ParsingException("I dont know what to do.");
            }
            root = new Binary(root, t, term());
        }

        return root;
    }

    @SuppressWarnings("unchecked")
    private AbstractSyntaxSubtree term()
    {
        AbstractSyntaxSubtree res = factor();
        while (_currToken.category() == TokenCategory.MULT ||
               _currToken.category() == TokenCategory.DIV)
        {
            Token<?> t = _currToken;
            switch (t.category())
            {
                case MULT:
                    eat(TokenCategory.MULT);
                    break;
                case DIV:
                    eat(TokenCategory.DIV);
                    break;
                default:
                    throw new ParsingException("Error when term-ing.");
            }
            res = new Binary(res, t, factor());
        }

        return res;
    }

    @SuppressWarnings("unchecked")
    private AbstractSyntaxSubtree factor()
    {
        Token<?> f = _currToken;
        AbstractSyntaxSubtree node;
        switch (f.category())
        {
            case PLUS:
                eat(TokenCategory.PLUS);
                return new Unary(f, factor());
            case MINUS:
                eat(TokenCategory.MINUS);
                return new Unary(f, factor());
            case INTEGER:
                eat(TokenCategory.INTEGER);
                return new NumberNode(f);
            case LPARENS:
                eat(TokenCategory.LPARENS);
                node = expression();
                eat(TokenCategory.RPARENS);
                return node;
            default:
                throw new ParsingException("Error when factorizing.");

        }
    }

    private void eat(final TokenCategory category)
    {
        if (_currToken.category() == category)
            _currToken = _lexer.next();
        else
            throw new ParsingException("Could not eat text.");
    }
}

File: src/parser.cpp
#include "parser.hpp"
#include "binary.hpp"
#include "unary.hpp"
#include "number_node.hpp"

void parser::init(const std::string& text) const
{
    lexer_.init(text);
    token_ = lexer_.next();
}

std::unique_ptr<ast> parser::parse() const
{
    return expression();
}

std::unique_ptr<ast> parser::expression() const
{
    std::unique_ptr<ast> curr(term());

    while (token_.category() == token_category::PLUS ||
           token_.category() == token_category::MINUS)
    {
        token t = token_;
        switch (t.category())
        {
            case token_category::PLUS:
                eat(token_category::PLUS);
                break;
            case token_category::MINUS:
                eat(token_category::MINUS);
                break;
            default:
                throw parsing_exception("I dont know what to do.");
        }

        curr = std::unique_ptr<binary>(new binary(std::move(curr), term(), t));
    }

    return curr;
}

std::unique_ptr<ast> parser::term() const
{
    std::unique_ptr<ast> curr(factor());

    while (token_.category() == token_category::MULT ||
           token_.category() == token_category::DIV)
    {
        token t = token_;
        switch (t.category())
        {
            case token_category::MULT:
                eat(token_category::MULT);
                break;
            case token_category::DIV:
                eat(token_category::DIV);
                break;
            default:
                throw parsing_exception("Error when term-ing.");
        }
        curr =
          std::unique_ptr<binary>(new binary(std::move(curr), factor(), t));
    }

    return curr;
}

std::unique_ptr<ast> parser::factor() const
{
    token f = token_;
    std::unique_ptr<ast> node;
    switch (f.category())
    {
        case token_category::PLUS:
            eat(token_category::PLUS);
            return std::unique_ptr<unary>(new unary(factor(), f));
        case token_category::MINUS:
            eat(token_category::MINUS);
            return std::unique_ptr<unary>(new unary(factor(), f));
        case token_category::INTEGER:
            eat(token_category::INTEGER);
            return std::unique_ptr<number_node>(new number_node(f));
        case token_category::LPARENS:
            eat(token_category::LPARENS);
            node = std::unique_ptr<ast>(expression());
            eat(token_category::RPARENS);
            return node;
        default:
            throw parsing_exception("Error when factorizing.");
    }
}

void parser::eat(const token_category& category) const
{
    if (token_.category() == category)
        token_ = lexer_.next();
    else
        throw parsing_exception("Could not eat text.");
}
